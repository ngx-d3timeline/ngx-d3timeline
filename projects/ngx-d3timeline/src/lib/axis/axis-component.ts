import { Component, Input } from '@angular/core';
import { Axis } from './axis';

@Component({
  selector: '[ngx-d3timeline-axis]',
  template: `
    <svg:g class="axis-group">
      <svg:g
        *ngFor="let tickMark of axis.tickMarks"
        ngx-d3timeline-axis-tick-mark
        [tickMark]="tickMark"
      ></svg:g>
      <svg:g ngx-d3timeline-line [line]="axis.line"></svg:g>
    </svg:g>
  `
})
export class AxisComponent {
  @Input() axis: Axis;
}
