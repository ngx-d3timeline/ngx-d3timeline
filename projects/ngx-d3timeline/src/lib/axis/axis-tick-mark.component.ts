import { Component, Input } from '@angular/core';
import { TickMark } from './tick-mark';

@Component({
  selector: '[ngx-d3timeline-axis-tick-mark]',
  template: `
    <svg:g [attr.transform]="tickMark.transform">
      <svg:text
        [attr.dx]="tickMark.labelOffset.x"
        [attr.dy]="tickMark.labelOffset.y"
      >
        {{ tickMark.label }}
      </svg:text>
      <svg:g
        ngx-d3timeline-line
        *ngIf="tickMark.line"
        [line]="tickMark.line"
      ></svg:g>
    </svg:g>
  `
})
export class AxisTickMarkComponent {
  @Input() tickMark: TickMark;
}
