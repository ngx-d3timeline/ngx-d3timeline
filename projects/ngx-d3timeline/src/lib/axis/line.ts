import { Orientation } from '../orientation';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export function createLine(
  x1: number,
  y1: number,
  length: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? createVerticalLine(x1, y1, length)
    : createHorizontalLine(x1, y1, length);
}

export function createVerticalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1,
    y2: y1 + length
  };
}

export function createHorizontalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1 + length,
    y2: y1
  };
}
