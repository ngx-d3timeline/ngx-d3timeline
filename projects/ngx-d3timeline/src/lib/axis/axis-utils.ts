import { Point } from '../core/point';
import { Scale } from '../scale-types';
import { Orientation } from '../core/orientation';
import { createLine, Line } from '../core/line';
import { Axis } from './axis';
import { TickMarkRenderer } from '../tick-mark/tick-mark-renderer';
import { getTickMarks } from '../tick-mark/tick-mark-utils';

function getRangeLimit(scale: Scale): number {
  return scale.range()[1];
}

function getAxisEndPoint(
  orientation: Orientation,
  scale: Scale,
  viewTopLeft: Point
): Point {
  return orientation === Orientation.Vertical
    ? { ...viewTopLeft, y: getRangeLimit(scale) }
    : { ...viewTopLeft, x: getRangeLimit(scale) };
}

function getAxisLine(
  viewTopLeft: Point,
  orientation: Orientation,
  scale: Scale
): Line {
  return createLine(
    viewTopLeft,
    getAxisEndPoint(orientation, scale, viewTopLeft)
  );
}

export function getAxis(
  getTickMarkRenderer: (
    scale: Scale,
    orientation: Orientation
  ) => TickMarkRenderer,
  scale: Scale,
  orientation: Orientation,
  viewTopLeft: Point
): Axis {
  return {
    line: getAxisLine(viewTopLeft, orientation, scale),
    tickMarks: getTickMarks(
      viewTopLeft,
      getTickMarkRenderer(scale, orientation)
    )
  };
}
