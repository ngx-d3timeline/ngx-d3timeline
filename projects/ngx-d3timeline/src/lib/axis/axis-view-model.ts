import { TickInfo } from './tick-info';
import { Line } from './line';

export interface AxisViewModel {
  tickInfos: TickInfo[];
  axisLine: Line;
}
