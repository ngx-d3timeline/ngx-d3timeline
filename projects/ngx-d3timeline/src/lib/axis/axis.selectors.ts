import { createSelector } from '../store-lib/selector/create-selector';
import {
  getAxis,
  getAxisLine,
  getAxisEndPoint,
  getTickGridLine
} from './axis-utils';
import { selectViewTopLeft } from '../view/view.selectors';
import {
  selectResourceAxisTickMarks,
  selectTimeAxisTickMarks,
  selectTimeAxisTickValues,
  selectResourceAxisTickValues,
  selectTickMarkTopLeftFunc
} from '../tick-mark/tick-mark.selector';
import { Line } from '../core/line';
import {
  selectOrientedBandScale,
  selectOrientedTimeScale
} from '../scales/scale-selectors';
import { OrientedScale } from '../scales/oriented-scale';
import { Scale } from '../scales/scale-types';
import { mapValues } from '../core/transform-utils';

const selectAxisLine = createSelector(selectViewTopLeft, viewTopLeft =>
  getAxisLine.bind(null, viewTopLeft)
);

const selectResourceAxisTickValueToGridLineFunc = createSelector(
  selectTickMarkTopLeftFunc,
  selectOrientedBandScale,
  selectOrientedTimeScale,
  (tickMarkTopLeft, orientedScale, otherOrientedScale) =>
    getTickGridLine.bind(
      null,
      tickMarkTopLeft,
      orientedScale,
      otherOrientedScale
    )
);

const selectTimeAxisTickValueToGridLineFunc = createSelector(
  selectTickMarkTopLeftFunc,
  selectOrientedTimeScale,
  selectOrientedBandScale,
  (tickMarkTopLeft, orientedScale, otherOrientedScale) =>
    getTickGridLine.bind(
      null,
      tickMarkTopLeft,
      orientedScale,
      otherOrientedScale
    )
);

export const selectResourceAxisGridLines = createSelector(
  selectResourceAxisTickValues,
  selectResourceAxisTickValueToGridLineFunc,
  mapValues
);

const selectTimeAxisGridLines = createSelector(
  selectTimeAxisTickValues,
  selectTimeAxisTickValueToGridLineFunc,
  mapValues
);

const selectResourceAxisLine = createSelector(
  selectOrientedBandScale,
  selectAxisLine,
  axisLineFromOrientedScale
);

export const selectTimeAxisLine = createSelector(
  selectOrientedTimeScale,
  selectAxisLine,
  axisLineFromOrientedScale
);

export const selectResourceAxis = createSelector(
  selectResourceAxisLine,
  selectResourceAxisTickMarks,
  selectResourceAxisGridLines,
  getAxis
);

export const selectTimeAxis = createSelector(
  selectTimeAxisLine,
  selectTimeAxisTickMarks,
  selectTimeAxisGridLines,
  getAxis
);

function axisLineFromOrientedScale(
  orientedScale: OrientedScale<Scale>,
  line: (o: OrientedScale<Scale>) => Line
) {
  return line(orientedScale);
}
