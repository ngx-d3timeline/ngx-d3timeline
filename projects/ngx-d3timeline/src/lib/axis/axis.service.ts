import { Axis } from './axis';
import { Orientation } from '../orientation';
import { map } from 'rxjs/operators';
import { Injectable } from '@angular/core';
import { TickMark } from './tick-mark';
import { Line } from './line';
import { Store } from '../store/store';
import { OptionsService } from '../options.service';
import { TimelineView } from '../view/timeline-view';
import { Scale } from '../scale-types';
import { ResourceAxisTickMarkRenderer } from './resources-axis/resource-axis-tick-mark-renderer';
import { TimeAxisTickMarkRenderer } from './time-axis/time-axis-tick-mark-renderer';
import { TickMarkRenderer } from './tick-mark-renderer';
import { flipOrientation } from '../orientation-utils';

@Injectable({ providedIn: 'root' })
export class AxisService {
  resourceAxis$ = this.store.state$.pipe(
    map(state =>
      this.createAxis(
        new ResourceAxisTickMarkRenderer(),
        state.bandScale,
        state.axisOrientations.resource,
        state.view
      )
    )
  );

  timeAxis$ = this.store.state$.pipe(
    map(state =>
      this.createAxis(
        new TimeAxisTickMarkRenderer(),
        state.timeScale,
        state.axisOrientations.time,
        state.view
      )
    )
  );

  constructor(private store: Store, private optionsService: OptionsService) {}

  private createAxis<TScale extends Scale>(
    tickMarkRenderer: TickMarkRenderer<TScale>,
    scale: TScale,
    orientation: Orientation,
    timelineView: TimelineView
  ): Axis {
    return {
      tickMarks: this.getTickMarks(
        tickMarkRenderer,
        scale,
        orientation,
        timelineView
      ),
      line: this.getLine(scale, orientation, timelineView)
    };
  }

  private getTickMarks<TScale extends Scale>(
    tickMarkRenderer: TickMarkRenderer<TScale>,
    scale: TScale,
    orientation: Orientation,
    timelineView: TimelineView
  ): TickMark[] {
    return tickMarkRenderer.getTickValues(scale).map(value => ({
      label: tickMarkRenderer.getTickLabel(scale, value),
      transform: this.optionsService.getTranslation(
        tickMarkRenderer.getTransform(scale, value),
        orientation,
        timelineView
      ),
      line: this.optionsService.getTickLine(
        tickMarkRenderer.getTickLineOffset(),
        flipOrientation(orientation)
      )
    }));
  }

  private getLine<TScale extends Scale>(
    scale: TScale,
    orientation: Orientation,
    timelineView: TimelineView
  ): Line {
    const line: Line = {
      x1: timelineView.left,
      x2: timelineView.left,
      y1: timelineView.top,
      y2: timelineView.top
    };
    const rangeLimit = scale.range()[1];

    return orientation === Orientation.Vertical
      ? { ...line, y2: rangeLimit }
      : { ...line, x2: rangeLimit };
  }
}
