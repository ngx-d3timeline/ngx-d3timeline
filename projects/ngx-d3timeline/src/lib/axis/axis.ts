import { TickMark } from '../tick-mark/tick-mark';
import { Line } from './line';

export interface Axis {
  tickMarks: TickMark[];
  line: Line;
}
