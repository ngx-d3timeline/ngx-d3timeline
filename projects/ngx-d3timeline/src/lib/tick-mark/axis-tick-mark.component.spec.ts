import { TestBed, ComponentFixture } from '@angular/core/testing';
import { AxisTickMarkComponent } from './axis-tick-mark.component';
import { createLine, Line } from '../axis/line';
import { Component, Input } from '@angular/core';

@Component({
  selector: '[ngx-d3timeline-line]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeLineComponent {
  @Input() line: Line;
}

describe('AxisTickMarkComponent', () => {
  let fixture: ComponentFixture<AxisTickMarkComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [AxisTickMarkComponent, FakeLineComponent]
    });

    fixture = TestBed.createComponent(AxisTickMarkComponent);
  });

  it('should not render line when tickMark.line is falsy', () => {
    fixture.componentInstance.tickMark = {
      label: 'tickMark label',
      labelOffset: { x: 0, y: 1 },
      line: null,
      transform: 'translate(10,10)'
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render line when tickMark.line is truthy', () => {
    fixture.componentInstance.tickMark = {
      label: 'tickMark label',
      labelOffset: { x: 0, y: 2 },
      line: createLine({ x: 0, y: 0 }, { x: 10, y: 10 }),
      transform: 'translate(10,10)'
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
