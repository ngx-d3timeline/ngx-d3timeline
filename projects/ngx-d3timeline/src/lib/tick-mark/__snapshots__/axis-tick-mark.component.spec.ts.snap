// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxisTickMarkComponent should not render line when tickMark.line is falsy 1`] = `
<div
  id="root0"
  ngx-d3timeline-axis-tick-mark=""
>
  <g
    transform="translate(10,10)"
  >
    <text
      dx="0"
      dy="1"
    >
       tickMark label 
    </text>
    
  </g>
</div>
`;

exports[`AxisTickMarkComponent should render line when tickMark.line is truthy 1`] = `
<div
  id="root1"
  ngx-d3timeline-axis-tick-mark=""
>
  <g
    transform="translate(10,10)"
  >
    <text
      dx="0"
      dy="2"
    >
       tickMark label 
    </text>
    
    <g
      class="tick-mark-line"
      ngx-d3timeline-line=""
    >
      <g />
    </g>
  </g>
</div>
`;
